arrows/ffmpeg/ffmpeg_video_input.cxx

```diff
@@ -0,0 +1,834 @@
+          &this->f_packet);
+        if (err == AVERROR_INVALIDDATA)
+        {// Ignore the frame and move to the next
+          av_free_packet(&this->f_packet);
+          continue;
+        }
+        if (err < 0)
+        {
+          LOG_ERROR(this->logger, "vidl_ffmpeg_istream: Error decoding packet");
+          av_free_packet(&this->f_packet);
+          return false;
+        }
+
+        this->f_pts = av_frame_get_best_effort_timestamp(this->f_frame);
+        if (this->f_pts == AV_NOPTS_VALUE)
+        {
+          this->f_pts = 0;
+        }
+      }
+      // \todo - No metadata support yet
+      //// grab the metadata from this packet if from the metadata stream
+      //else if (this->packet.stream_index == this->data_index)
+      //{
+      //  is_->metadata_.insert(is_->metadata_.end(), is_->packet_.data,
+      //    is_->packet_.data + is_->packet_.size);
+      //}
+
+      if (!got_picture)
+      {
+        av_free_packet(&this->f_packet);
+      }
+    }
+
+    // From ffmpeg apiexample.c: some codecs, such as MPEG, transmit the
+    // I and P frame with a latency of one frame. You must do the
+    // following to have a chance to get the last frame of the video.
+    if (!got_picture)
+    {
+      av_init_packet(&this->f_packet);
+      this->f_packet.data = nullptr;
+      this->f_packet.size = 0;
+
+      int err = avcodec_decode_video2(this->f_video_encoding,
+        this->f_frame, &got_picture,
+        &this->f_packet);
+      if (err >= 0)
+      {
+        this->f_pts += static_cast<int64_t>(this->stream_time_base_to_frame());
+      }
+    }
+
+    // The cached frame is out of date, whether we managed to get a new
+    // frame or not.
+    this->current_image_memory = nullptr;
+
+    if (!got_picture)
+    {
+      this->f_frame->data[0] = NULL;
+    }
+
+    return got_picture != 0;
+  }
+
+  // ==================================================================
+  /*
+  * @brief Get the current timestamp
+  *
+  * @return \b Current timestamp.
+  */
+  double current_pts() const
+  {
+    return this->f_pts * av_q2d(this->f_video_stream->time_base);
+  }
+
+  // ==================================================================
+  /*
+  * @brief Returns the double value to convert from a stream time base to
+  *  a frame number
+  */
+  double stream_time_base_to_frame() const
+  {
+    if (this->f_video_stream->avg_frame_rate.num == 0.0)
+    {
+      return av_q2d(av_inv_q(av_mul_q(this->f_video_stream->time_base,
+        this->f_video_stream->r_frame_rate)));
+    }
+    return av_q2d(
+      av_inv_q(
+        av_mul_q(this->f_video_stream->time_base, this->f_video_stream->avg_frame_rate)));
+  }
+
+  bool is_valid() const
+  {
+    return this->f_frame && this->f_frame->data[0];
+  }
+
+  // ==================================================================
+  /*
+  * @brief Return the current frame number
+  *
+  * @return \b Current frame number.
+  */
+  unsigned int frame_number() const
```

Comment:
same comment as above.  We should keep a frame counter (how many times you advanced) and return that.  This gets tricky when dealing with seeking.  It may require a novel solution to make seeking reliably frame accurate.

Reply:
Again, I would remove this function.  It has been shown to be broken in VXL and we don't use it in `vidl_ffmpeg_video_input`.  We remember the frame number in a member variable instead.